architecture/2024-07-user-defined-types.md

```diff
@@ -0,0 +1,329 @@
+
+When users interact with Radius through the CLI, include sample 
+input commands and their corresponding output. Include a bicep/helm code 
+sample, if this proposal involves updates to that experience.
+-->
+
+**Sample Input:**
+<!--
+Provide a sample CLI command input and/or bicep/helm code.
+-->
+
+**Sample Output:**
+<!--
+Provide a sample output for the inputs provided above.
+-->
+
+**Sample Recipe Contract:**
+<!--
+Provide a sample of updated recipe contract, if this proposal 
+updates recipe contract (input parameters, outputs schema, etc)
+-->
+
+## Design
+
+### High Level Design
+
+User-defined types functionality can be broken down into a few different dataflows and systems.
+
+- **Registration and management of user-defined types**: This is a new API that provides CRUDL behavior for the concept of *resource type*.
+- **Request path for user-defined types**: This includes routing, validation, storage, and versioning. We typically refer to this today in Radius as the *frontend*.
+- **Processing for user-defined types**: This includes Recipes or any other asynchronous processing that occurs as a result of a `PUT` or `DELETE` operation. We typically refer to this today in Radius as the *async-worker*.
+- **Tooling that interacts with the Radius API**: This includes the `rad` CLI and Dashboard. When the set of resource-types in Radius is dynamic, then the tools need to be dynamic as well. 
+- **Tooling to author user-defined type schemas**: This includes any tooling we produce to validate a schema, or to transform a schema into a Bicep extension. 
+
+We can also break these systems down by their components in Radius. 
+- Registration and management of user-defined types is part of UCP. Since UCP is the *front-controller* for Radius and is responsible for routing, it makes logical sense that UCP would implement the management of user-defined types.
+- Request path for user-defined types is either part of UCP or a new component. 
+	- UCP implements routing, and will validate that the incoming request is destined for a valid resource type. 
+	- The other *frontend* concerns (validation, storage, versioning) could either be implemented inside UCP or another new component. 
+- Processing for user-defined types is either part of UCP or a new component.
+	- The async-worker for user-defined types should probably be hosted in the same component as the `frontend` like our existing resource providers.
+- Tooling that interacts with the Radius API is outside the control plane. These features are implemented in the relevant tool. 
+- Tooling to author user-defined type schemas is outside the control plane. Depending on the exact needs this is part of the `rad` CLI or some other tool. 
+
+From this can extract the idea of a *UDT RP* - this is a new component that executes the declarative logic of a user-defined type. It's like our existing resource providers, except it's dynamic and driven by the schema definitions provided by users.
```

Comment:
Were you thinking of moving existing namespaces to this RP as well?

Reply:
As much as possible. I think being able to use UDT for **most** of our existing resources is a success criteria. 

Keyword most. Resource types like `secretStores` are special.